### PR TITLE
Escaping and unescaping of non-printable characters in strings

### LINKED ISSUE #79: Multi-line strings
What is the format for multiline strings? Is:
```
"one line
another line","another attribute"
```

correct?

Or

```
"one line\
another line","another attribute"
```
?

Or

```
"one line\nanother line","another attribute"
```
?

Or any of the above??

### LINKED ISSUE #89: Issues when string data has a character `\n` or `\r\n`
I am not sure that this is something easy to handle.
Currently, if the data contain a string with the escape character `\n` or`\r\n`, then the parser will fail. I would assume this due to the fact that those characters are used to split the data as well.

### PR DESCRIPTION
**Pull request by jnothman**
fixes #89
Fixes #79

### CODE REVIEW

**Comment by jnothman** [comment]
Implementation complete (?). @mfeurer, want to update docs for me? :\

**Comment by mfeurer** [comment]
Hey @jnothman thanks a lot for the fast implementation, but I'm sorry, I'll be out of office for a week and won't be able to check this before. I will check this afterwards.

**Comment by jnothman** [comment]
That's alright. No hurry. I also have weird legacy test cases to work out how to support.

**Comment by jnothman** [comment]
Py2 problems...

**Comment by jnothman** [comment]
Green!

File: arff.py [hunk 1]
```diff
@@ -223,9 +223,40 @@ def _build_re_values():
...
+    '\\n': '\n',
+    '\\r': '\r',
+    '\\b': '\b',
+    '\\f': '\f',
+    '\\%': '%',
+}
+_UNESCAPE_SUB_MAP = {chr(i): '\\%03o' % i for i in range(32)}
+_UNESCAPE_SUB_MAP.update({v: k for k, v in _ESCAPE_SUB_MAP.items()})
+_UNESCAPE_SUB_MAP[''] = '\\'
+_ESCAPE_SUB_MAP.update({'\\%d' % i: chr(i) for i in range(10)})
```
[the remark this answers is not in the document]
Comment (jnothman) [reply arff.py]: In the callback below you can see the logic where `_ESCAPE_SUB_MAP` is only used in the case of a single-character escape

**Comment by jnothman** [comment]
What are the differences from weka? Yes, for several characters there are multiple choices of representation in encoding.

**Comment by mfeurer** [comment]
Reading the file
```
@RELATION bla
@ATTRIBUTE attr STRING
@DATA
'a,b,c'
'a,b,c '
'a\\\\c'
'a\\"c'
'a\\'c'
'a\\nc'
'a\\tc'
'a\\000c'
'a\\017c'
```
with WEKA 3.8 and then saving it again results in
```
@relation bla

@attribute attr string

@data
'a,b,c'
'a,b,c '
'a\\\\c'
'a\\\"c'
'a\\'
''
'a\\nc'
'a\\tc'
'a\\000c'
'a\\017c'
```
while executing the following python script
```
import arff

ARFF = """@RELATION bla
@ATTRIBUTE attr STRING
@DATA
'a,b,c'
'a,b,c '
'a\\\\c'
'a\\"c'
'a\\'c'
'a\\nc'
'a\\tc'
'a\\000c'
'a\\017c'
"""

data = arff.loads(ARFF, return_type=arff.DENSE)
with open('tmp.arff', 'wt') as fh:
    fh.write(arff.dumps(data))
```
results in
```
@RELATION bla

@ATTRIBUTE attr STRING

@DATA
'a,b,c'
'a,b,c '
'a\\c'
'a\"c'
'a\'c'
'a\nc'
'a\tc'
'a\000c'
'a\017c'
```

**Comment by jnothman** [comment]
You should be *reading* the file in Python too. You've entered the string incorrectly for comparison. You should be using `r"""` for instance.

**Comment by jnothman** [comment]
Ahh. I see now. The ARFF that's being loaded in the test is:

```
@RELATION bla
 
 @ATTRIBUTE attr STRING
 
@DATA
'a,b,c'
'a,b,c '
'a\\c'
'a\"c'
'a\'c'
'a\nc'
'a\tc'
'a\000c'
'a\017c'
```

**Comment by jnothman** [comment]
Merge and release, @mfeurer?

**Comment by mfeurer** [comment]
Would after the ICML deadline be okay?

**Comment by jnothman** [comment]
Of course!

**Comment by jnothman** [comment]
How about now? ;)

**Comment by mfeurer** [comment]
Sorry for the delay, it was indeed still on my stack. This looks good to me. I'll do a release now, and @arlindkadra will update the documentation in the near future.

**Comment by mfeurer** [comment]
It's live now: https://pypi.org/project/liac-arff/

**Comment by jnothman** [comment]
Great, thanks!

**Pull request merged**